catalog/src/main/java/org/killbill/billing/catalog/io/VersionedCatalogLoader.java

```diff
@@ -116,36 +122,46 @@ private URL getURLFromString(final String urlString) {
         try {
-            final DefaultVersionedCatalog result = new DefaultVersionedCatalog();
+            final Collection<Future<StandaloneCatalog>> catalogs = new ArrayList<Future<StandaloneCatalog>>(catalogXMLs.size());
             for (final String cur : catalogXMLs) {
-                final InputStream curCatalogStream = new ByteArrayInputStream(cur.getBytes());
-                final StandaloneCatalog catalog = XMLLoader.getObjectFromStream(curCatalogStream, StandaloneCatalog.class);
-                if (!filterTemplateCatalog || !catalog.isTemplateCatalog()) {
-                    result.add(new StandaloneCatalogWithPriceOverride(catalog, priceOverride, tenantRecordId, internalCallContextFactory));
+                catalogs.add(executorService.submit(new Callable<StandaloneCatalog>() {
+
+                    @Override
+                    public StandaloneCatalog call() throws Exception {
+                        final InputStream curCatalogStream = new ByteArrayInputStream(cur.getBytes());
+                        final StandaloneCatalog catalog = XMLLoader.getObjectFromStream(curCatalogStream, StandaloneCatalog.class);
+                        if (!filterTemplateCatalog || !catalog.isTemplateCatalog()) {
+                            return new StandaloneCatalogWithPriceOverride(catalog, priceOverride, tenantRecordId, internalCallContextFactory);
+                        }
+                        return null;
+                    }
+                }));
+            }
+
+            final DefaultVersionedCatalog result = new DefaultVersionedCatalog();
+            for (final Future<StandaloneCatalog> standaloneCatalogFuture : catalogs) {
+                final StandaloneCatalog catalog = standaloneCatalogFuture.get();
+                if (catalog != null) {
+                    result.add(catalog);
                 }
             }
+
             XMLLoader.initializeAndValidate(result);
             return result;
         } catch (final ValidationException e) {
             logger.warn("Failed to load catalog for tenantRecordId='{}'", tenantRecordId, e);
-            for (ValidationError ve : e.getErrors()) {
+            for (final ValidationError ve : e.getErrors()) {
                 logger.warn(ve.toString());
             }
             throw new CatalogApiException(e, ErrorCode.CAT_INVALID_FOR_TENANT, tenantRecordId);
-        } catch (final JAXBException e) {
+        } catch (final InterruptedException e) {
             logger.warn("Failed to load catalog for tenantRecordId='{}'", tenantRecordId, e);
             throw new CatalogApiException(e, ErrorCode.CAT_INVALID_FOR_TENANT, tenantRecordId);
-        } catch (final IOException e) {
-            logger.warn("Failed to load catalog for tenantRecordId='{}'", tenantRecordId, e);
-            throw new IllegalStateException(e);
-        } catch (final TransformerException e) {
+        } catch (final ExecutionException e) {
```

Comment:
I thought about it, but since we re-wrap it in a `CatalogApiException` anyways, I figured dropping one level of indirection would have been confusing.